File: questions/104.MaximumDepthOfBinaryTree.go
package questions

import "math"

func maxDepth(root *TreeNode) int {
	if root == nil {
		return 0
	}

	left := maxDepth(root.Left)
	right := maxDepth(root.Right)

	return 1 + int(math.Max(float64(left), float64(right)))
}

File: leetcode/104.MaximumDepthOfBinaryTree.ts
import { TreeNode } from './shared'

function maxDepth(root: TreeNode | null): number {
  const dfs = (node: TreeNode | null): number => {
    if (!node) return 0

    const left = dfs(node.left)
    const right = dfs(node.right)

    return 1 + Math.max(left, right)
  }

  return dfs(root)
}

const tree = new TreeNode(3)
tree.left = new TreeNode(9)
tree.right = new TreeNode(20)
tree.right.left = new TreeNode(15)
tree.right.right = new TreeNode(7)

console.log(maxDepth(tree))
